solidity/contracts/bridge/VendingMachine.sol

```diff
@@ -0,0 +1,307 @@
+        _mint(msg.sender, amount);
+    }
+
+    /// @notice Mints TBTC v2 to `from` address from TBTC v1 with 1:1 ratio.
+    ///         `from` address needs to have at least `amount` of TBTC v1
+    ///         balance approved for transfer to the `VendingMachine` before
+    ///         calling this function.
+    /// @dev This function is a shortcut for approve + mint. Only TBTC v1
+    ///      caller is allowed and only TBTC v1 is allowed as a token to
+    ///      transfer.
+    /// @param from TBTC v1 token holder minting TBTC v2 tokens
+    /// @param amount The amount of TBTC v2 to mint from TBTC v1
+    /// @param token TBTC v1 token address
+    function receiveApproval(
+        address from,
+        uint256 amount,
+        address token,
+        bytes calldata
+    ) external override {
+        require(token == address(tbtcV1), "Token is not TBTC v1");
+        require(msg.sender == address(tbtcV1), "Only TBTC v1 caller allowed");
+        _mint(from, amount);
+    }
+
+    /// @notice Unmints TBTC v2 from the caller into TBTC v1. Depending on
+    ///         `unmintFee` value, may require paying an additional unmint fee
+    ///         in TBTC v2 in addition to the amount being unminted. To see
+    ///         what is the value of the fee, please call `unmintFeeFor(amount)`
+    ///         function. The caller needs to have at least
+    ///         `amount + unmintFeeFor(amount)` of TBTC v2 balance approved for
+    ///         transfer to the `VendingMachine` before calling this function.
+    /// @param amount The amount of TBTC v2 to unmint to TBTC v1
+    function unmint(uint256 amount) external {
+        uint256 fee = unmintFeeFor(amount);
+        emit Unminted(msg.sender, amount, fee);
+
+        require(
+            tbtcV2.balanceOf(msg.sender) >= amount + fee,
+            "Amount + fee exceeds TBTC v2 balance"
+        );
+
+        tbtcV2.safeTransferFrom(msg.sender, address(this), fee);
+        tbtcV2.burnFrom(msg.sender, amount);
+        tbtcV1.safeTransfer(msg.sender, amount);
+    }
+
+    /// @notice Allows the Governance to withdraw unmint fees accumulated by
+    ///         `VendingMachine`.
+    /// @param recipient The address receiving the fees
+    /// @param amount The amount of fees in TBTC v2 to withdraw
+    function withdrawFees(address recipient, uint256 amount)
+        external
+        onlyOwner
+    {
+        tbtcV2.safeTransfer(recipient, amount);
+    }
+
+    /// @notice Initiates unmint fee update process. The update process needs to
+    ///         be finalized with a call to `finalizeUnmintFeeUpdate` function
+    ///         after the `GOVERNANCE_DELAY` passes. Only unmint fee update
+    ///         initiator role can initiate the update.
+    /// @param _newUnmintFee The new unmint fee
+    function initiateUnmintFeeUpdate(uint256 _newUnmintFee)
+        external
+        only(unmintFeeUpdateInitiator)
+    {
+        /* solhint-disable-next-line not-rely-on-time */
```

Comment:
Do we want to check if the `_newUnmintFee` isn't exceeding 100%?

Reply:
But maybe it's not necessary as we've got ACL.

Reply:
Yeah, I think we can leave it as-is given there is an ACL here.